### DISCUSSION TITLE
Shrine 2 support for Ruby 3 AND/OR Shrine 3 :versions_compatibility flag

### DISCUSSION THREAD

**Discussion by rinconchris** [shrinerb/shrine#615] [Q&A] [answered]
I'm in the process of updating a large rails app to ruby 3, and am trying to find the best way to handle Shrine. Currently using Shrine 2.19.0 and very happy with it - so would prefer not to upgrade Shrine if that's possible. 

Can anyone confirm if there's a way to get Shrine 2 running with Ruby 3? I've seen some discussion about resolving the ruby 3 argument issues, but I think they were all for Shrine 3. 

We've also written a number of custom uploaders (for direct to S3) based on the shrine 2 storage format with an image_data column containing an :original and a :thumb key. I briefly experimented with Shrine 3 using the `versions_compatibility: true` option to see how it would work, and unfortunately found that the the image_url(:type) now seems to ignore the param and always return the url for the :original version. Any thoughts on a workaround for this?

**Comment by benkoshy** [comment]
> Can anyone confirm if there's a way to get Shrine 2 running with Ruby 3? I've seen some discussion about resolving the ruby 3 argument issues, but I think they were all for Shrine 3.

The only way I can think of is to run Shrine 2 tests using Ruby 3. You may do this locally on a computer, or you may use github's CI:

- Here is the [Shrine CI workflow](https://github.com/shrinerb/shrine/blob/master/.github/workflows/ci.yml).
- Consider checking out the version of Shrine you want (Shrine 2), and also set the ruby matrix to the version you require (3.1 etc.)  and see if the tests pass, you can do this in your own fork of the repository.

I cannot comment on the second part of your question - where the param is ignored etc.

**Comment by janko** [comment, marked as answer]
You should be able to just use the `versions` plugin on Shrine 3. It is deprecated, but I highly doubt there will ever be a Shrine 4, so it will technically never be removed. Would this work for you?

**Reply by rinconchris** [reply]
Thanks guys for the quick responses! I was able to get things going using Shrine 3.4 along with the :processing and :versions plugins instead of :derivatives. Only other change needed for Ruby 3 compatibility was to update the arguments in generate_location method to add the double-spat ** `generate_location(io, **context)`
